src/iop/primaries.c

```diff
@@ -0,0 +1,382 @@
+{
+  dt_iop_module_t *self = (dt_iop_module_t *)user_data;
+  gui_changed(self, NULL, NULL);
+}
+
+static GtkWidget *_setup_hue_slider(dt_iop_module_t *self, const char *param_name, const char *tooltip)
+{
+  GtkWidget *slider = dt_bauhaus_slider_from_params(self, param_name);
+  dt_bauhaus_slider_set_format(slider, "°");
+  dt_bauhaus_slider_set_digits(slider, 1);
+  dt_bauhaus_slider_set_factor(slider, RAD_TO_DEG);
+  dt_bauhaus_slider_set_soft_range(slider, -20.f / RAD_TO_DEG, 20.f / RAD_TO_DEG);
+  gtk_widget_set_tooltip_text(slider, tooltip);
+  return slider;
+}
+
+static GtkWidget *_setup_purity_slider(dt_iop_module_t *self, const char *param_name, const char *tooltip)
+{
+  GtkWidget *slider = dt_bauhaus_slider_from_params(self, param_name);
+  dt_bauhaus_slider_set_format(slider, "%");
+  dt_bauhaus_slider_set_digits(slider, 1);
+  dt_bauhaus_slider_set_factor(slider, 100.f);
+  dt_bauhaus_slider_set_offset(slider, -100.f);
+  dt_bauhaus_slider_set_soft_range(slider, 0.5f, 1.5f);
+  gtk_widget_set_tooltip_text(slider, tooltip);
+  return slider;
+}
+
+void gui_init(dt_iop_module_t *self)
+{
+  dt_iop_primaries_gui_data_t *g = IOP_GUI_ALLOC(primaries);
+
+  g->red_hue = _setup_hue_slider(self, "red_hue", _("red primary hue"));
+  g->red_purity = _setup_purity_slider(self, "red_purity", _("red primary purity"));
+  g->green_hue = _setup_hue_slider(self, "green_hue", _("green primary hue"));
+  g->green_purity = _setup_purity_slider(self, "green_purity", _("green primary purity"));
+  g->blue_hue = _setup_hue_slider(self, "blue_hue", _("blue primary hue"));
+  g->blue_purity = _setup_purity_slider(self, "blue_purity", _("blue primary purity"));
+
+  g->achromatic_tint_hue = dt_bauhaus_slider_from_params(self, "achromatic_tint_hue");
+  dt_bauhaus_slider_set_format(g->achromatic_tint_hue, "°");
+  dt_bauhaus_slider_set_digits(g->achromatic_tint_hue, 1);
+  dt_bauhaus_slider_set_factor(g->achromatic_tint_hue, RAD_TO_DEG);
+  gtk_widget_set_tooltip_text(g->achromatic_tint_hue, _("tint hue"));
+
+  g->achromatic_tint_purity = dt_bauhaus_slider_from_params(self, "achromatic_tint_purity");
+  dt_bauhaus_slider_set_format(g->achromatic_tint_purity, "%");
+  dt_bauhaus_slider_set_digits(g->achromatic_tint_purity, 1);
+  dt_bauhaus_slider_set_factor(g->achromatic_tint_purity, 100.f);
+  dt_bauhaus_slider_set_soft_range(g->achromatic_tint_purity, 0.f, 0.2f);
+  gtk_widget_set_tooltip_text(g->achromatic_tint_purity, _("tint purity"));
+
+  g->painted_work_profile = NULL;
+  g->painted_display_profile = NULL;
+
+  DT_DEBUG_CONTROL_SIGNAL_CONNECT(darktable.signals, DT_SIGNAL_CONTROL_PROFILE_USER_CHANGED,
+                                  G_CALLBACK(_signal_profile_user_changed), self);
+  DT_DEBUG_CONTROL_SIGNAL_CONNECT(darktable.signals, DT_SIGNAL_CONTROL_PROFILE_CHANGED,
+                                  G_CALLBACK(_signal_profile_changed), self);
+  DT_DEBUG_CONTROL_SIGNAL_CONNECT(darktable.signals, DT_SIGNAL_DEVELOP_UI_PIPE_FINISHED,
```

Comment:
Hi @dterrahe, thanks for notifying. However, I can't reproduce the crash the way you suggested.

The code uses the same handler for both `DT_SIGNAL_CONTROL_PROFILE_CHANGED` and `DT_SIGNAL_DEVELOP_UI_PIPE_FINISHED`, and that handler (`_signal_profile_changed`) is already disconnected in `gui_cleanup()`. The disconnect function seems to use internally this one:
https://docs.gtk.org/gobject/func.signal_handlers_disconnect_matched.html
and it seems to disconnect every signal where the handler and data match.

Not quite sure how to proceed with this. Do you have a way to reproduce this and get a stack trace?

Reply:
You are absolutely right!

I guess I got confused because I more recently looked at `g_idle_remove_by_data` and `g_source_remove_by_user_data` which only remove one handler, so one might have to loop (until it returns FALSE) if there are possibly multiple handlers (or just call it twice if you know there are two, as would have been the case here). The `g_signal_handlers_` family, as you point out, works on _all_ connected handlers.

I did test when I reported this and I did get a crash, but it must have been caused by something else and I can't reproduce now. Sorry for wasting your time.

As an aside; I did consider at one point whether `default_gui_cleanup` should be extended to do a `g_signal_handlers_disconnect_by_data(darktable.signals, self);` so that most explicit `gui_cleanup` implementations could be removed, because mostly (not exclusively!) they are used to disconnect signals. The drawback would be that we would spend some time trying to find connected signals for modules that don't have any and that people might connect with other data than the module itself (not a problem; it would just still require an explicit `gui_cleanup` to disconnect of which there would then be fewer examples).

Do you think this would be a useful simplication? It could done at the same time as replacing all
`DT_DEBUG_CONTROL_SIGNAL_CONNECT(darktable.signals, `
calls with 
`DT_CONTROL_SIGNAL_CONNECT(` 
(i.e. hide the implementation detail that these signals are connected to the `darktable.signals` object or that (dis)connections can be debugged). 
With same for `DT_DEBUG_CONTROL_SIGNAL_DISCONNECT(darktable.signals, ` -> `DT_CONTROL_SIGNAL_DISCONNECT(` 
and 
`DT_DEBUG_CONTROL_SIGNAL_RAISE(darktable.signals, ` -> `DT_CONTROL_SIGNAL_RAISE(`. 

Since `DT_CONTROL_SIGNAL_CONNECT` would still be a macro, it can also take care of casting the handler by wrapping it in a `G_CALLBACK()`, so we could simplify
```
  DT_DEBUG_CONTROL_SIGNAL_CONNECT(darktable.signals, DT_SIGNAL_CONTROL_PROFILE_USER_CHANGED,
                                  G_CALLBACK(_signal_profile_user_changed), self);
```
to
```
  DT_CONTROL_SIGNAL_CONNECT(DT_SIGNAL_CONTROL_PROFILE_USER_CHANGED, _signal_profile_user_changed, self);
```
In my view more readable/obvious with less scary magic.
@TurboGit @ralfbrown what do you think?

Reply:
> I did test when I reported this and I did get a crash, but it must have been caused by something else and I can't reproduce now. Sorry for wasting your time.

No problem. Glad to see someone spending effort testing and reporting things.

Also nice suggestions on improving things around signals. My two cents (keep in mind I'm not very deep into the codebase, so just commenting by gut feeling):

> people might connect with other data than the module itself (not a problem; it would just still require an explicit gui_cleanup to disconnect of which there would then be fewer examples)

The cost of this would be increase cognitive load to those authoring or reading module code. A somewhat obscure rule to keep in mind: "if signal handlers with data other than `self` are connected, they must be disconnected in an explicit `gui_cleanup()`." This feels somewhat magical and could cause confusion. I think, in this case, explicit is better than implicit.

On the other hand, a quick grep through the repo shows that all IOPs have either `self` or some value derived from it (e.g. `self->dev`) as user data. Perhaps one could go as far as add a new macro `DT_IOP_CONTROL_SIGNAL_CONNECT()` which would automatically pass `self` as user data, and this would be used in all IOPs to connect signals? Then we could promise to automatically disconnect all the signals added by that, without an explicit `gui_cleanup()`. Not sure if this can be done elegantly.

> `DT_CONTROL_SIGNAL_CONNECT(DT_SIGNAL_CONTROL_PROFILE_USER_CHANGED, _signal_profile_user_changed, self);`

Yeah, this makes sense to me. Much cleaner and does away with the GLib details.

Reply:
> this would be increase cognitive load to those authoring or reading module code

Agreed. That's why I left it at the time, I guess. We _could_ in theory add a call to `g_signal_handler_find` after `gui_cleanup`, only in debug mode, to check if any handlers were left connected inadvertently. That might help finding sources of crashes with less effort. But many people never compile in debug mode (it didn't use to work reliably).

> or some value derived from it (e.g. `self->dev`) as user data

Only once and can easily be rewritten.

> Not sure if this can be done elegantly.

Depends on whether you find picking up a local variable by name within a macro without taking it as a parameter "elegant". We could also do the next best thing and typecheck the `user_data` in the call to `dt_iop_module_t *`. But tbh introducing a separate `_IOP_` version and expecting people to understand that in _other_ places they should disconnect a non-iop one... I'm probably tempted to drop that again.

> does away with the GLib details

To me making things short so that it's easier to see the relevant details is worth it most of the time. I hate boiler plate, since I'm not paid per line of code...